---
title: "EDGER pipeline to analyse RNAseq data from QUANTseq seed"
author: "Hélène Zuber"
date: "06022019"
output:
  html_document: 
  toc: yes
html_notebook:
  number_sections: yes
theme: journal
toc: yes
toc_depth: 5
editor_options: 
  chunk_output_type: console
---

In this script version, I analyze results from script-seq library. New analysis from 15012019. Statistical analysis with Wt, urt1, xrn4 and urt1xrn4
```{r "setup", include=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR")# Setting working directory
```
/Users/helenez/Documents/divers_work/QUANTSEQ2018/scriptseq_2014/EdgeR
### Load required package
```{R}
require(plyr)
require(dplyr)
library(edgeR)
library(xlsx)
```

### Import and formating of the data set
**Import of the data**
```{R}
#give directory with htseq-count files
filenames <- list.files("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/htseqcount/collapse", full.names=TRUE)
filenames_1 <- filenames[c(1,7,13, 2, 8, 14, 3, 9, 15, 4, 10, 16, 5, 11, 17, 6, 12, 18)] # reorder the order of import 
#create a list of dataframe with htseq-count results
File_list <- lapply(filenames_1, read.delim, header=FALSE)
filenames_1
```

**Merge all data together**
```{R}
# give names to the dataframe within the list
names(File_list) <- c("Col0_FDS_rep1", "Col0_FDS_rep2", "Col0_FDS_rep3", 
                      "Col0_20_rep1",  "Col0_20_rep2", "Col0_20_rep3",
                      "Col0_25_rep1",  "Col0_25_rep2", "Col0_25_rep3",
                      "urt1_FDS_rep1", "urt1_FDS_rep2", "urt1_FDS_rep3", 
                      "urt1_20_rep1",  "urt1_20_rep2", "urt1_20_rep3",
                      "urt1_25_rep1",  "urt1_25_rep2", "urt1_25_rep3"
                      )

#merge all tables together
htcount <- Reduce(function(...) left_join(..., by = "V1"), File_list)
htcount_names <- c("AGI", names(File_list))
colnames(htcount) <-htcount_names

#remove unuseful lines at the end of the htseq count (check that these the right light to be removed)
htcount<-htcount[-c(37337:37342),]
```

**Format table in order to create EdgeR object**
```{R}
count <- htcount[ ,2:19]
rownames( count ) <- htcount[,1]
#give library sizes
colSums( count ) # Library Sizes
table( rowSums( count ) )[ 1:30 ] # Number of genes with low counts
#indicates different conditions and replicates
condition <- gl(6,3,labels=c("Col0_FDS", "Col0_20", "Col0_25", "urt1_FDS", "urt1_20", "urt1_25")) # 6 different conditions for 3 replicates
replicate <- factor(rep(c("rep1","rep2", "rep3"),6)) # 3 different replicates for 6 conditions
```


### Building the edgeR Object
```{R}
cds <- DGEList( count , group = condition)
names( cds )
head(cds$counts) # original count matrix
cds$samples # contains a summary of your samples

#remove AGI with low read number. Here I keep only those genes that have at least 1 read per 10000 (around 10 reads for theses libraries) in at least 2 sample.

keep <- rowSums(cpm(cds)>1) >= 2
cds <- cds[keep, , keep.lib.sizes=FALSE]
#calculate normalisation factor
cds <- calcNormFactors(cds)
cds$samples #give Normalisation factor
#method="TMM" is the weighted trimmed mean of M-values (to the reference) proposed by #Robinson and Oshlack (2010), where the weights are from the delta method on Binomial #data. If refColumn is unspecified, the library whose upper quartile is closest to the #mean upper quartile is used.

cds$samples$lib.size * cds$samples$norm.factors #to give effective library size
```


```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR")
## equation from the edgeR documentation for estimating normalized absolute expression from their scaling factors
ScaleFactors <- cds$samples$lib.size * cds$samples$norm.factors
Exp <- round(t(t(cds$counts)/ScaleFactors) * mean(ScaleFactors)) # standard to round b/c can't have a fraction of a read
list_universe <- as.vector(row.names(Exp)) # list that will be used for GO analysis latter
head(cpm(cds))
write.table(Exp, "Norm_data_total.txt", sep="\t", col.names = NA)
write.table(cpm(cds), "CPM.txt", sep="\t", col.names = NA)
```

## Multi-Dimensional Scaling Plot analysis to vizualize distance between samples
```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/MDS")
myPalette <- rep(c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "red", "green"),each=3) ##color blind colors
top = nrow(cds[[1]])
# Output plot as a pdf
pdf( "MDS_plot_all.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds , top = top, main = "MDS Plot for all conditions", labels = colnames( cds$counts ),col=myPalette)
dev.off() # this tells [R] to close and stop writing to the pdf.
pdf( "MDS_plot_defaut.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds , main = "MDS Plot for all conditions", labels = colnames( cds$counts ),col=myPalette)
dev.off() # this tells [R] to close and stop writing to the pdf.
##alternative method
pdf( "MDS_plot_bcv.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds , main = "MDS Plot for all conditions", method = "bcv", labels = colnames( cds$counts ),col=myPalette)
dev.off()
```

### Make different plots for different tissue samples (FDS, 20, 25)
```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/MDS")
myPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "red", "green") ##color blind colors

cds_FDS <- cds[,c(1,2,3,10,11,12)]
cds_20 <- cds[,c(4,5,6,13,14,15)]
cds_25 <- cds[,c(7,8,9,16,17,18)]

# Output plot as a pdf
##alternative method
pdf( "MDS_plot_bcv_cds_FDS.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds_FDS , main = "MDS Plot for all conditions", method = "bcv", labels = colnames( cds_FDS$counts ),col=myPalette)
dev.off()

pdf( "MDS_plot_bcv_cds_20.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds_20 , main = "MDS Plot for all conditions", method = "bcv", labels = colnames( cds_20$counts ),col=myPalette)
dev.off()

pdf( "MDS_plot_bcv_cds_25.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds_25 , main = "MDS Plot for all conditions", method = "bcv", labels = colnames( cds_25$counts ),col=myPalette)
dev.off()


pdf( "MDS_plot_defaut_cds_FDS.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds_FDS , main = "MDS Plot for all conditions", labels = colnames( cds_FDS$counts ),col=myPalette)
dev.off()

pdf( "MDS_plot_defaut_cds_20.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds_20 , main = "MDS Plot for all conditions", labels = colnames( cds_20$counts ),col=myPalette)
dev.off()

pdf( "MDS_plot_defaut_cds_25.pdf" , width = 7 , height = 7 ) # in inches
plotMDS( cds_25 , main = "MDS Plot for all conditions", labels = colnames( cds_25$counts ),col=myPalette)
dev.off()
```

## Differential gene analysis

**Make the design for the GLM approach and give contrasts**
```{R}
design <- model.matrix(~0+group+replicate,data=cds$samples)
design
```
Contrasts for statistical analysis (paiwise analysis)
```{R}
my.contrasts <-makeContrasts(wtvsurt1_FDS= groupurt1_FDS-groupCol0_FDS,
                             wtvsurt1_20= groupurt1_20-groupCol0_20, 
                             wtvsurt1_25= groupurt1_25-groupCol0_25,
                             
                             FDSvs20_wt= groupCol0_20-groupCol0_FDS,
                             FDSvs25_wt= groupCol0_25-groupCol0_FDS,
                             c20vs25_wt= groupCol0_25-groupCol0_20,
                             
                             FDSvs20_urt1= groupurt1_20-groupurt1_FDS,
                             FDSvs25_urt1= groupurt1_25-groupurt1_FDS,
                             c20vs25_urt1= groupurt1_25-groupurt1_20,
                             
                             levels=design)
my.contrasts
```

## Statistical test to compare genotypes
```{R}
dge = estimateGLMCommonDisp(cds, design)
dge = estimateGLMTagwiseDisp(dge, design) #each gene gets assigned the same dispersion estimate
# Fit data to the glm model
fit <- glmFit(dge,design)
```

```{R 'annotation function'}
library("biomaRt")
addBiomaRtAnnotation <- function(normObj, biomart="plants_mart",
    dataset="athaliana_eg_gene", host="plants.ensembl.org",
    features=c("ensembl_peptide_id","ensembl_transcript_id","ensembl_gene_id",
    "external_gene_name"))#, GO=FALSE)
{
    stopifnot(!missing(normObj))
    stopifnot()
    filter <- match.arg(features)
    mart <- useMart(biomart,dataset,host=host)
    atr <- c(features,"description")
    geneInfo <- getBM(rownames(normObj),mart = mart, attributes = atr,filters = filter) 
    if (length(geneInfo$description)==0) {stop(sprintf("rowIDs from count table are different than IDs from \"%s\" from ensembl\nChoose another feature ID, or another dataset from biomaRt, or check that your IDs are in the format as the one from ensembl", filter))}
    print(length(geneInfo$description))
    print(length(row.names(normObj)))
    rownames(geneInfo) <- geneInfo[,1] # C'est toujours la premiere colonne qui contient les identifiants.
    normObj <- data.frame(normObj)
    normObj$id  <- 1:nrow(normObj)
    normObj <- merge(normObj, geneInfo, by="row.names", all.x=TRUE)
    rownames(normObj) <- normObj$Row.names
    normObj$Row.names <- NULL
    normObj <-normObj[order(normObj$id), ]
     normObj$id <- NULL
    normObj
}
```



```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/list")
# Pairwise comparison (by defaut use genewise dispersion)
my_test <- function(A) {
  rowtoprint <- as.data.frame(data.frame(comparison=as.character(), up_genes=as.numeric(), down_genes=as.numeric()) )          
  x <- glmLRT(fit, contrast=my.contrasts[,A])
  y<-topTags(x, n = nrow(x$table))$table
  table_up <- subset(y, y$FDR<0.05&y$logFC>0)
  table_up <- try(addBiomaRtAnnotation(table_up, features="ensembl_gene_id"))
  table_down <- subset(y, y$FDR<0.05&y$logFC<0)
  table_down <- try(addBiomaRtAnnotation(table_down, features="ensembl_gene_id"))
  #print(paste("up", nrow(table_up)), sep="\t")
  #print(paste("down", nrow(table_down)), sep="\t")
  rowtoprint <- as.data.frame(data.frame(comparison=as.character(A), up_genes=as.numeric(nrow(table_up)), down_genes=as.numeric(nrow(table_down)))) 
  write.table(table_up, paste(i, "up.txt", sep="_"), quote=FALSE, sep="\t", col.names=NA)
  write.table(table_down, paste(i, "down.txt", sep="_"), quote=FALSE, sep="\t", col.names=NA)
  return(rowtoprint)
  
  }


recap <- as.data.frame(data.frame(comparison=as.character(), up_genes=as.numeric(), down_genes=as.numeric()) )          
for (i in colnames(my.contrasts)) {
                                 print(i)
                                 tab <- my_test(i)
                                 recap <- as.data.frame(rbind(recap, tab))
                                 colnames(recap) <- c("comparison", "up_genes", "down_genes")
                                  }
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR")
write.table(recap, "recap_number_withbatch.txt", quote=FALSE, sep="\t", col.names=NA)
 
```

```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/list0.01")
# Pairwise comparison (by defaut use genewise dispersion)
my_test <- function(A) {
  rowtoprint <- as.data.frame(data.frame(comparison=as.character(), up_genes=as.numeric(), down_genes=as.numeric()) )          
  x <- glmLRT(fit, contrast=my.contrasts[,A])
  y<-topTags(x, n = nrow(x$table))$table
  table_up <- subset(y, y$FDR<0.01&y$logFC>0)
  table_up <- try(addBiomaRtAnnotation(table_up, features="ensembl_gene_id"))
  table_down <- subset(y, y$FDR<0.01&y$logFC<0)
  table_down <- try(addBiomaRtAnnotation(table_down, features="ensembl_gene_id"))
  #print(paste("up", nrow(table_up)), sep="\t")
  #print(paste("down", nrow(table_down)), sep="\t")
  rowtoprint <- as.data.frame(data.frame(comparison=as.character(A), up_genes=as.numeric(nrow(table_up)), down_genes=as.numeric(nrow(table_down)))) 
  write.table(table_up, paste(i, "up.txt", sep="_"), quote=FALSE, sep="\t", col.names=NA)
  write.table(table_down, paste(i, "down.txt", sep="_"), quote=FALSE, sep="\t", col.names=NA)
  return(rowtoprint)
  
  }


recap <- as.data.frame(data.frame(comparison=as.character(), up_genes=as.numeric(), down_genes=as.numeric()) )          
for (i in colnames(my.contrasts)) {
                                 print(i)
                                 tab <- my_test(i)
                                 recap <- as.data.frame(rbind(recap, tab))
                                 colnames(recap) <- c("comparison", "up_genes", "down_genes")
                                  }
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR")
write.table(recap, "recap_number_withbatch_0.01.txt", quote=FALSE, sep="\t", col.names=NA)
 
```

## GO analysis (using GOstats package)

```{R}
library("GOstats")
library(siggenes)
library(EMA)
library(fdrtool) # package to correct p-value
library("org.At.tair.db") # package with Arabidopsis annotation
library("GO.db")
library(DBI)
```


```{R}
gene_list <- c()
type <- 'BP'
my_param <- function(gene_list,type) new('GOHyperGParams', geneIds=gene_list, universeGeneIds=list_universe,
              ontology=type,
              pvalueCutoff=1,
              conditional=T,
              testDirection='over',
              annotation="org.At.tair.db")

```



```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/GO")
my_GO <- function(A) {
  x <- glmLRT(fit, contrast=my.contrasts[,A])
  y<-topTags(x, n = nrow(x$table))$table
  table_up <- subset(y, y$FDR<0.05&y$logFC>0)
  table_down <- subset(y, y$FDR<0.05&y$logFC<0)
  #up
  my_over_test <- hyperGTest(my_param(as.vector(row.names(table_up)), 'BP'))
  r_BP <- summary(my_over_test)
  r_BP$type <- "BP"
  colnames(r_BP) <- c("GOID", "Pvalue", "OddsRatio",	"ExpCount",	"Count",	"Size",	"Term", "Type")
  my_over_test <- hyperGTest(my_param(as.vector(row.names(table_up)), 'MF'))
  r_MF <- summary(my_over_test)
  r_MF$type <- "MF"
  colnames(r_MF) <- c("GOID", "Pvalue", "OddsRatio",	"ExpCount",	"Count",	"Size",	"Term", "Type")
  my_over_test <- hyperGTest(my_param(as.vector(row.names(table_up)), 'CC'))
  r_CC <- summary(my_over_test)
  r_CC$type <- "CC"
  colnames(r_CC) <- c("GOID", "Pvalue", "OddsRatio",	"ExpCount",	"Count",	"Size",	"Term", "Type")
  r <- rbind(r_BP, r_MF, r_CC)
  pval <- r$Pvalue
  adjpvalsK <- multiple.correction(pval,typeFDR="FDR-BH")
  fdr = fdrtool(pval, statistic="pvalue",plot=FALSE)
  r[,9]<-as.data.frame(adjpvalsK)
  r[,10]<-as.data.frame(fdr$qval)  
  r <- r[order(adjpvalsK),]
  write.table(r, paste(i, "up.txt", sep="_"), quote=FALSE, sep="\t", col.names=NA)

  ##down
 my_over_test <- hyperGTest(my_param(as.vector(row.names(table_down)), 'BP'))
  r_BP <- summary(my_over_test)
  r_BP$type <- "BP"
  colnames(r_BP) <- c("GOID", "Pvalue", "OddsRatio",	"ExpCount",	"Count",	"Size",	"Term", "Type")
  my_over_test <- hyperGTest(my_param(as.vector(row.names(table_down)), 'MF'))
  r_MF <- summary(my_over_test)
  r_MF$type <- "MF"
  colnames(r_MF) <- c("GOID", "Pvalue", "OddsRatio",	"ExpCount",	"Count",	"Size",	"Term", "Type")
  my_over_test <- hyperGTest(my_param(as.vector(row.names(table_down)), 'CC'))
  r_CC <- summary(my_over_test)
  r_CC$type <- "CC"
  colnames(r_CC) <- c("GOID", "Pvalue", "OddsRatio",	"ExpCount",	"Count",	"Size",	"Term", "Type")
  r <- rbind(r_BP, r_MF, r_CC)
  pval <- r$Pvalue
  adjpvalsK <- multiple.correction(pval,typeFDR="FDR-BH")
  fdr = fdrtool(pval, statistic="pvalue",plot=FALSE)
  r[,9]<-as.data.frame(adjpvalsK)
  r[,10]<-as.data.frame(fdr$qval) 
  r <- r[order(adjpvalsK),]
  write.table(r, paste(i, "down.txt", sep="_"), quote=FALSE, sep="\t", col.names=NA)
  }



for (i in colnames(my.contrasts)) {
                                 print(i)
                                 my_GO(i)
                                  }

```

### Venn diagram
```{R}
#give directory with the list of deregulated genes
name_list <- list.files("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/list", full.names=TRUE)

comparison_genotype <- name_list[c(17,18,13,14,15,16)]
name_list_genotype <- lapply(comparison_genotype, read.delim, header=TRUE)
names(name_list_genotype) <- c("wtvsurt1_FDS_down", "wtvsurt1_FDS_up", "wtvsurt1_20_down", "wtvsurt1_20_up", "wtvsurt1_20_down",  "wtvsurt1_25_up")


comparison_FDSvs20 <- name_list[c(7, 8, 5, 6)]
comparison_FDSvs25 <- name_list[c(11, 12, 9, 10)]
comparison_20vs25 <- name_list[c(3, 4, 1, 2)]

name_list_FDSvs20 <- lapply(comparison_FDSvs20, read.delim, header=TRUE)
name_list_FDSvs25 <- lapply(comparison_FDSvs25, read.delim, header=TRUE)
name_list_20vs25 <- lapply(comparison_20vs25, read.delim, header=TRUE)

names(name_list_FDSvs20) <- c("wt_down", "wt_up", "urt1_down", "urt1_up")
names(name_list_FDSvs25) <- c("wt_down", "wt_up", "urt1_down", "urt1_up")
names(name_list_20vs25) <- c("wt_down", "wt_up", "urt1_down", "urt1_up")
```


# Venn diagram to compare genes misregulated in urt1 in the different conditions
```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/Venn")
library("VennDiagram")
my_color <- c("darkmagenta", "darkblue",  "darkorange")
my_category <- c("FDS","20", "25")
# draw diagram Venn
my_diag <- function(x, z) {
                          list_genes <- lapply(x, function(y) y[,1])                        
                          geneLS_down <- list_genes[c(1,3,5)]
                          geneLS_up <- list_genes[c(2,4,6)]
                          names(geneLS_down) <- my_category
                          names(geneLS_up) <- my_category
                          pdf( paste(z, "vennDiagramm_up.pdf") , width = 7 , height = 7,pointsize = 30)
                          venn.plot <- venn.diagram(geneLS_up , NULL, 
                                                    fill=my_color, 
                                                    alpha=c(0.5,0.5,0.5), cex = 0.5, cat.fontface=3, cat.cex=0.8,  
                                                    category.names=my_category, 
                                                    main="Upregulated genes")
                                                    # To plot the venn diagram we will use the   grid.draw() function to plot the venn diagram
                                                    grid.draw(venn.plot)
                        dev.off()
                        pdf( paste(z, "vennDiagramm_down.pdf") , width = 7 , height = 7,pointsize = 30)
                          venn.plot <- venn.diagram(geneLS_down , NULL, 
                                                    fill=my_color, 
                                                    alpha=c(0.5,0.5,0.5), cex = 0.5, cat.fontface=3, cat.cex=0.8,  
                                                    category.names=my_category, 
                                                    main="Downregulated genes")
                                                    # To plot the venn diagram we will use the   grid.draw() function to plot the venn diagram
                                                    grid.draw(venn.plot)
                        dev.off()
                        
}

my_diag(name_list_genotype, "genotype")

```


# Venn diagram to compare genes misregulated between the different conditions for WT and urt1
```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/Venn")
library("VennDiagram")
my_color <- c("darkgreen", "darkred")
my_category <- c("wt","urt1")
# draw diagram Venn
my_diag <- function(x, z) {
                          list_genes <- lapply(x, function(y) y[,1])                        
                          geneLS_down <- list_genes[c(1,3)]
                          geneLS_up <- list_genes[c(2,4)]
                          names(geneLS_down) <- my_category
                          names(geneLS_up) <- my_category
                          pdf( paste(z, "vennDiagramm_up.pdf") , width = 7 , height = 7,pointsize = 30)
                          venn.plot <- venn.diagram(geneLS_up , NULL, 
                                                    fill=my_color, 
                                                    alpha=c(0.5,0.5), cex = 0.5, cat.fontface=3, cat.cex=0.8,  
                                                    category.names=my_category, 
                                                    main="Upregulated genes")
                                                    # To plot the venn diagram we will use the   grid.draw() function to plot the venn diagram
                                                    grid.draw(venn.plot)
                        dev.off()
                        pdf( paste(z, "vennDiagramm_down.pdf") , width = 7 , height = 7,pointsize = 30)
                          venn.plot <- venn.diagram(geneLS_down , NULL, 
                                                    fill=my_color, 
                                                    alpha=c(0.5,0.5), cex = 0.5, cat.fontface=3, cat.cex=0.8,  
                                                    category.names=my_category, 
                                                    main="Downregulated genes")
                                                    # To plot the venn diagram we will use the   grid.draw() function to plot the venn diagram
                                                    grid.draw(venn.plot)
                        dev.off()
                        
}

my_diag(name_list_FDSvs20, "FDSvs20")
my_diag(name_list_FDSvs25, "FDSvs25")
my_diag(name_list_20vs25, "20vs25")
```

**Import of the GO data**
```{R}
#give directory with htseq-count files
GO_list <- list.files("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/GO", full.names=TRUE)
GO_list_down <- GO_list[c(seq(1, 17, by = 2))]
GO_list_up <- GO_list[c(seq(2, 18, by = 2))]

GO_list_down_1 <- GO_list_down[c(9,7,8, 4, 3, 6, 5, 2, 1)]
GO_list_up_1 <- GO_list_up[c(9,7,8, 4, 3, 6, 5, 2, 1)]

#create a list of dataframe with htseq-count results
GO_down <- lapply(GO_list_down_1, read.delim, header=TRUE)
GO_up <- lapply(GO_list_up_1, read.delim, header=TRUE)

names(GO_down) <- c("A_wtvsurt1_FDS", "B_wtvsurt1_20", "C_wtvsurt1_25", "D_FDS_20_wt", "E_FDS_20_urt1", "F_FDS_25_wt", "G_FDS_25_urt1", "H_20_25_wt", "I_20_25_urt1")
names(GO_up) <- c("A_wtvsurt1_FDS", "B_wtvsurt1_20", "C_wtvsurt1_25", "D_FDS_20_wt", "E_FDS_20_urt1", "F_FDS_25_wt", "G_FDS_25_urt1", "H_20_25_wt", "I_20_25_urt1")
```

```{r}
##function to add column with name of the file in a supplemental column
my_add_geno <-  function(x)
{
  for (i in names(x)) 
  { 
    print(i)
    x[[i]]$comparison <- i
  }
  return(x)
}
```

```{r}
GO_down_1 <- my_add_geno(GO_down)
GO_up_1 <- my_add_geno(GO_up)
#GO_down_2 <- lapply(GO_down_1, function(x) x[x$adjpvalsK<0.0001,])
#GO_up_2 <- lapply(GO_up_1, function(x) x[x$adjpvalsK<0.0001,])
```

#Make list for GO to keep 
```{R}
GO_down_3 <- lapply(GO_down_1, function(x) x[,c(2,10)])
GO_down_list<- Reduce(function(dtf1, dtf2) merge(dtf1, dtf2, by = "GOID", all = TRUE),  GO_down_3)
GO_down_list[is.na(GO_down_list)] <- 1
row.names(GO_down_list) <- GO_down_list$GOID
GO_down_list <- GO_down_list[,-1]
keep <- rowSums(GO_down_list<0.00000001) >= 1 #(1E-8)
GO_down_selected <- as.vector(row.names(GO_down_list[keep,]))


GO_up_3 <- lapply(GO_up_1, function(x) x[,c(2,10)])
GO_up_list<- Reduce(function(dtf1, dtf2) merge(dtf1, dtf2, by = "GOID", all = TRUE),  GO_up_3)
GO_up_list[is.na(GO_up_list)] <- 1
row.names(GO_up_list) <- GO_up_list$GOID
GO_up_list <- GO_up_list[,-1]
keep <- rowSums(GO_up_list<0.00000001) >= 1 #(1E-8)
GO_up_selected <- as.vector(row.names(GO_up_list[keep,]))
```



```{R}
#GO_down_table <- Reduce(function(dtf1, dtf2) merge(dtf1, dtf2, by = "GOID", all = TRUE),  GO_down_2)
GO_down_table <- do.call("rbind", GO_down_1)
GO_up_table <- do.call("rbind", GO_up_1)
GO_down_table <- GO_down_table[,-1]
GO_up_table <- GO_up_table[,-1]

GO_down_table <- GO_down_table[GO_down_table$GOID %in% GO_down_selected,]
GO_up_table <- GO_up_table[GO_up_table$GOID %in% GO_up_selected,]
```

```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/")
library(ggplot2)

GO_down_table$sign <- as.factor((GO_down_table$adjpvalsK<0.05))
p <- ggplot(GO_down_table) + geom_point(aes(y=Count, x=reorder(paste(GOID, substr(Term, 1, 30)), -Count), fill=sign), size = 3, shape=22) +
      facet_wrap("comparison", ncol=1, scales="free_y") + theme(panel.spacing = unit(0.4, "lines")) +
                    theme ( plot.title = element_text(hjust = 0.5),
                    #panel.grid.major = element_blank(), 
                    #panel.grid.minor = element_blank(),
                    #panel.background = element_blank(), 
                    panel.border = element_rect(colour="black",fill=NA,size=0.5),
                    #axis.line = element_line(colour = "black"),
                    #legend.position="none",
                    axis.text.x = element_text(size=8, angle=90, hjust = 1),
                    strip.text.x = element_text(size=8, face="bold")) +
                    scale_fill_manual(values=c("black", "darkseagreen1"))
p

ggsave(filename = "GO_plot_down.pdf", p, width = 10, height = 20, dpi = 300)
```
 


```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/")

GO_up_table$sign <- as.factor((GO_up_table$adjpvalsK<0.05))
p <- ggplot(GO_up_table) + geom_point(aes(y=Count, x=reorder(paste(GOID, substr(Term, 1, 30)), -Count), fill=sign), size = 3, shape=22) +
      facet_wrap("comparison", ncol=1, scales="free_y") + theme(panel.spacing = unit(0.4, "lines")) +
                    theme ( plot.title = element_text(hjust = 0.5),
                    #panel.grid.major = element_blank(), 
                    #panel.grid.minor = element_blank(),
                    #panel.background = element_blank(), 
                    panel.border = element_rect(colour="black",fill=NA,size=0.5),
                    #axis.line = element_line(colour = "black"),
                    #legend.position="none",
                    axis.text.x = element_text(size=8, angle=90, hjust = 1),
                    strip.text.x = element_text(size=8, face="bold")) +
                    scale_fill_manual(values=c("black", "red3"))
                    
p

ggsave(filename = "GO_plot_up.pdf", p, width = 10, height = 20, dpi = 300)
```




**Import of the GO data bis**
```{R}
#give directory with htseq-count files
GO_list <- list.files("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/GO", full.names=TRUE)
GO_list_down <- GO_list[c(seq(1, 17, by = 2))]
GO_list_up <- GO_list[c(seq(2, 18, by = 2))]

GO_list_down_1 <- GO_list_down[c(9,7,8)]
GO_list_up_1 <- GO_list_up[c(9,7,8)]

#create a list of dataframe with htseq-count results
GO_down <- lapply(GO_list_down_1, read.delim, header=TRUE)
GO_up <- lapply(GO_list_up_1, read.delim, header=TRUE)

names(GO_down) <- c("A_wtvsurt1_FDS", "B_wtvsurt1_20", "C_wtvsurt1_25")
names(GO_up) <- c("A_wtvsurt1_FDS", "B_wtvsurt1_20", "C_wtvsurt1_25")
```

```{r}
##function to add column with name of the file in a supplemental column
my_add_geno <-  function(x)
{
  for (i in names(x)) 
  { 
    print(i)
    x[[i]]$comparison <- i
  }
  return(x)
}
```

```{r}
GO_down_1 <- my_add_geno(GO_down)
GO_up_1 <- my_add_geno(GO_up)
#GO_down_2 <- lapply(GO_down_1, function(x) x[x$adjpvalsK<0.0001,])
#GO_up_2 <- lapply(GO_up_1, function(x) x[x$adjpvalsK<0.0001,])
```

#Make list for GO to keep 
```{R}
GO_down_3 <- lapply(GO_down_1, function(x) x[,c(2,10)])
GO_down_list<- Reduce(function(dtf1, dtf2) merge(dtf1, dtf2, by = "GOID", all = TRUE),  GO_down_3)
GO_down_list[is.na(GO_down_list)] <- 1
row.names(GO_down_list) <- GO_down_list$GOID
GO_down_list <- GO_down_list[,-1]
keep <- rowSums(GO_down_list<0.001) >= 1 #(1E-8)
GO_down_selected <- as.vector(row.names(GO_down_list[keep,]))


GO_up_3 <- lapply(GO_up_1, function(x) x[,c(2,10)])
GO_up_list<- Reduce(function(dtf1, dtf2) merge(dtf1, dtf2, by = "GOID", all = TRUE),  GO_up_3)
GO_up_list[is.na(GO_up_list)] <- 1
row.names(GO_up_list) <- GO_up_list$GOID
GO_up_list <- GO_up_list[,-1]
keep <- rowSums(GO_up_list<0.001) >= 1 #(1E-8)
GO_up_selected <- as.vector(row.names(GO_up_list[keep,]))
```



```{R}
#GO_down_table <- Reduce(function(dtf1, dtf2) merge(dtf1, dtf2, by = "GOID", all = TRUE),  GO_down_2)
GO_down_table <- do.call("rbind", GO_down_1)
GO_up_table <- do.call("rbind", GO_up_1)
GO_down_table <- GO_down_table[,-1]
GO_up_table <- GO_up_table[,-1]

GO_down_table <- GO_down_table[GO_down_table$GOID %in% GO_down_selected,]
GO_up_table <- GO_up_table[GO_up_table$GOID %in% GO_up_selected,]
```

```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/")
library(ggplot2)

GO_down_table$sign <- as.factor((GO_down_table$adjpvalsK<0.05))
p <- ggplot(GO_down_table) + geom_point(aes(y=Count, x=reorder(paste(GOID, substr(Term, 1, 30)), -Count), fill=sign), size = 4, shape=22) +
      facet_wrap("comparison", ncol=1) + theme(panel.spacing = unit(0.4, "lines")) +
                    theme ( plot.title = element_text(hjust = 0.5),
                    #panel.grid.major = element_blank(), 
                    #panel.grid.minor = element_blank(),
                    #panel.background = element_blank(), 
                    panel.border = element_rect(colour="black",fill=NA,size=0.5),
                    #axis.line = element_line(colour = "black"),
                    #legend.position="none",
                    axis.text.x = element_text(size=12, angle=90, hjust = 1),
                    strip.text.x = element_text(size=8, face="bold")) +
                    scale_fill_manual(values=c("black", "darkseagreen1"))
p

ggsave(filename = "GO_plot_down_bis.pdf", p, width = 10, height = 10, dpi = 300)
```
 


```{R}
setwd("/Users/hzuber/Documents/RNAseq/Quant-seq2019/Analysis/EdgeR/")

GO_up_table$sign <- as.factor((GO_up_table$adjpvalsK<0.05))
p <- ggplot(GO_up_table) + geom_point(aes(y=Count, x=reorder(paste(GOID, substr(Term, 1, 30)), -Count), fill=sign), size = 3, shape=22) +
      facet_wrap("comparison", ncol=1) + theme(panel.spacing = unit(0.4, "lines")) +
                    theme ( plot.title = element_text(hjust = 0.5),
                    #panel.grid.major = element_blank(), 
                    #panel.grid.minor = element_blank(),
                    #panel.background = element_blank(), 
                    panel.border = element_rect(colour="black",fill=NA,size=0.5),
                    #axis.line = element_line(colour = "black"),
                    #legend.position="none",
                    axis.text.x = element_text(size=12, angle=90, hjust = 1),
                    strip.text.x = element_text(size=8, face="bold")) +
                    scale_fill_manual(values=c("black", "red3"))
                    
p

ggsave(filename = "GO_plot_up_bis.pdf", p, width = 10, height = 10, dpi = 300)
```


